packages/x-data-grid/src/components/containers/GridRootStyles.ts

```diff
@@ -273,6 +280,13 @@ export const GridRootStyles = styled('div', {
     [`& .${c.columnHeader}:focus, & .${c.cell}:focus`]: {
       outline: `solid ${t.palette.primary.main} 1px`,
     },
+    [`&.${c['root--noToolbar']} [aria-rowindex="1"] [aria-colindex="1"]`]: {
+      borderTopLeftRadius: 'var(--unstable_DataGrid-radius)',
+    },
+    [`&.${c['root--noToolbar']} [aria-rowindex="1"] [aria-colindex="${visibleColumns.length}"]`]: {
+      borderTopRightRadius:
+        'calc((1 - var(--DataGrid-hasScrollY)) * var(--unstable_DataGrid-radius))', // if (!hasScrollY) { style.borderTopRightRadius = radius }
+    },
```

Comment:
```suggestion
    [`&:not(:has(.${c.toolbarContainer})) [aria-rowindex="1"] [aria-colindex="1"]`]: {
      borderTopLeftRadius: 'var(--unstable_DataGrid-radius)',
    },
    [`&:not(:has(.${c.toolbarContainer}))[aria-rowindex="1"] [aria-colindex="${visibleColumns.length}"]`]:
      {
        borderTopRightRadius:
          'calc((1 - var(--DataGrid-hasScrollY)) * var(--unstable_DataGrid-radius))', // if (!hasScrollY) { style.borderTopRightRadius = radius }
      },
```

Could use a combination of `:not(:has())` here if it's preferable to adding a new class.

Reply:
I try to avoid non-class selectors when possible, more details in [these links](https://mui-org.slack.com/archives/C03SUL8TC1M/p1707306827856999). It's a balance to have with performance vs readability, but one important thing to note is that adding complex CSS selectors makes styling slower, because unlike JS which is run one line at a time, CSS selectors are always matched on all elements. Class selectors being the most common ones also means browsers optimize those heavily, but not so much other complex selectors.

In this particular case I'm not sure it would work though, the toolbar container is not in the hierarchy of the header cells IIRC.

Reply:
I tried it and it does work, but fair point on the performance and readability, it's probably not worth changing.